pkg/asset/machines/clusterapi.go

```diff
@@ -0,0 +1,175 @@
+	"k8s.io/utils/pointer"
+	capa "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/machines/aws"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	"github.com/openshift/installer/pkg/asset/rhcos"
+	awstypes "github.com/openshift/installer/pkg/types/aws"
+	awsdefaults "github.com/openshift/installer/pkg/types/aws/defaults"
+)
+
+func GenerateClusterAPI(ctx context.Context, installConfig *installconfig.InstallConfig, clusterID *installconfig.ClusterID, rhcosImage *rhcos.Image) (*capiutils.GenerateMachinesOutput, error) {
+	var err error
+	result := &capiutils.GenerateMachinesOutput{}
+	ic := installConfig.Config
+	pool := *ic.ControlPlane
+
+	switch ic.Platform.Name() {
+	case awstypes.Name:
+		subnets := map[string]string{}
+		if len(ic.Platform.AWS.Subnets) > 0 {
+			subnetMeta, err := installConfig.AWS.PrivateSubnets(ctx)
+			if err != nil {
+				return nil, err
+			}
+			for id, subnet := range subnetMeta {
+				subnets[subnet.Zone.Name] = id
+			}
+		}
+
+		mpool := defaultAWSMachinePoolPlatform("master")
+
+		osImage := strings.SplitN(string(*rhcosImage), ",", 2)
+		osImageID := osImage[0]
+		if len(osImage) == 2 {
+			osImageID = "" // the AMI will be generated later on
+		}
+		mpool.AMIID = osImageID
+
+		mpool.Set(ic.Platform.AWS.DefaultMachinePlatform)
+		mpool.Set(pool.Platform.AWS)
+		zoneDefaults := false
+		if len(mpool.Zones) == 0 {
+			if len(subnets) > 0 {
+				for zone := range subnets {
+					mpool.Zones = append(mpool.Zones, zone)
+				}
+			} else {
+				mpool.Zones, err = installConfig.AWS.AvailabilityZones(ctx)
+				if err != nil {
+					return nil, err
+				}
+				zoneDefaults = true
+			}
+		}
+
+		if mpool.InstanceType == "" {
+			topology := configv1.HighlyAvailableTopologyMode
+			if pool.Replicas != nil && *pool.Replicas == 1 {
+				topology = configv1.SingleReplicaTopologyMode
+			}
+			mpool.InstanceType, err = aws.PreferredInstanceType(ctx, installConfig.AWS, awsdefaults.InstanceTypes(installConfig.Config.Platform.AWS.Region, installConfig.Config.ControlPlane.Architecture, topology), mpool.Zones)
+			if err != nil {
+				logrus.Warn(errors.Wrap(err, "failed to find default instance type"))
+				mpool.InstanceType = awsdefaults.InstanceTypes(installConfig.Config.Platform.AWS.Region, installConfig.Config.ControlPlane.Architecture, topology)[0]
+			}
+		}
+
+		// if the list of zones is the default we need to try to filter the list in case there are some zones where the instance might not be available
+		if zoneDefaults {
+			mpool.Zones, err = aws.FilterZonesBasedOnInstanceType(ctx, installConfig.AWS, mpool.InstanceType, mpool.Zones)
+			if err != nil {
+				logrus.Warn(errors.Wrap(err, "failed to filter zone list"))
+			}
+		}
```

Comment:
I looked into deduplicating this logic. Let's just merge this PR and I will follow up with a PR--probably to convert machinepools to an asset.